File: Sesion 11/Marco teorico/9ARCHIVOS DE ACCESO ALEATORIO/5.py
import struct
class Persona:
    def __init__(self, nombre, edad):
        self.nombre = nombre
        self.edad = edad

def escribir_personas():
    personas = [
        Persona("Juan Perez", 30),
        Persona("Maria Gomez", 45),
        Persona("Carlos Ruiz", 35)
    ]

    with open('personas.dat', 'wb') as archivo_salida:
        for persona in personas:
            datos_persona = struct.pack('50s i', persona.nombre.encode('utf-8'), persona.edad)
            archivo_salida.write(datos_persona)

    print("Datos escritos en el archivo correctamente.")

def leer_persona(indice):
    try:
        with open('personas.dat', 'rb') as archivo_entrada:
            desplazamiento = indice * struct.calcsize('50s i')
            archivo_entrada.seek(desplazamiento)
            datos_persona = archivo_entrada.read(struct.calcsize('50s i'))
            nombre, edad = struct.unpack('50s i', datos_persona)
            nombre = nombre.decode('utf-8').strip('\x00')  # Decodificar y eliminar caracteres nulos
            print(f"Nombre: {nombre}, Edad: {edad}")
    except FileNotFoundError:
        print("No se pudo abrir el archivo para lectura.")


escribir_personas()

print("\nLeer persona en posición 0:")
leer_persona(0)

print("\nLeer persona en posición 1:")
leer_persona(1)

print("\nLeer persona en posición 2:")
leer_persona(2)

File: Sesion 11/Marco teorico/9ARCHIVOS DE ACCESO ALEATORIO/5.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>
using namespace std;


struct Persona {
    char nombre[50];
    int edad;
};


void escribirPersonas() {
    ofstream archivoSalida("personas.dat", ios::binary);


    if (!archivoSalida) {
        cerr << "No se pudo abrir el archivo para escritura." << endl;
        return;
    }


    Persona personas[] = {
        {"Juan Perez", 30},
        {"Maria Gomez", 45},
        {"Carlos Ruiz", 35}
    };


    for (const auto& persona : personas) {
        archivoSalida.write(reinterpret_cast<const char*>(&persona), sizeof(Persona));
    }


    archivoSalida.close();
}


void leerPersona(int indice) {
    ifstream archivoEntrada("personas.dat", ios::binary);


    if (!archivoEntrada) {
        cerr << "No se pudo abrir el archivo para lectura." << endl;
        return;
    }


    archivoEntrada.seekg(indice * sizeof(Persona), ios::beg);


    Persona persona;
    archivoEntrada.read(reinterpret_cast<char*>(&persona), sizeof(Persona));


    if (archivoEntrada) {
        cout << "Nombre: " << persona.nombre << ", Edad: " << persona.edad << endl;
    } else {
        cerr << "No se pudo leer la persona en la posición " << indice << endl;
    }


    archivoEntrada.close();
}


int main() {
    escribirPersonas();


    cout << "Leer persona en posición 0:" << endl;
    leerPersona(0);


    cout << "Leer persona en posición 1:" << endl;
    leerPersona(1);


    cout << "Leer persona en posición 2:" << endl;
    leerPersona(2);


    return 0;
}
